Make sure credits in the new subscription will cover costs generated by previous subscription that are terminated

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-17805

Review by commit 🐠 

Do not be afraid of the number of commits here, a lot of them are just refactoring commits and are easy to review.
The number of files changed looks intimidating is because I renamed a util function!

The issue is, the domain has subscription A, in the middle of the month, it is changed to subscription B. All credits for subscription A is also transferred to subscription B. At the end of the month, when generating invoices, all costs incurred from subscription B will be covered by those credits, but since those credits are tied to subscription B, subscription A will remain unpaid, which can lead to domain being paused when overdue.

We already have logic implemented in https://github.com/dimagi/commcare-hq/pull/26831 to make sure `ANY` type credit will cover terminated subscriptions in the same month. But we didn't have logic to make sure other type of credits will. So this PR is to make sure that the same logic will be applied to other type of credits.

Some context to review this PR:
What consists of the total of an invoice:
- Plan fee
- Feature fee ( User, SMS, ... )
Both plan fee and feature fee is considered as line items.

Types of Credit:
- ANY ( Can be used to pay any fee left remained in the total)
- Product ( Can only be used to pay Plan fee)
- Feature ( Can only be used to pay Feature fee)


## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

Added new tests in `corehq/apps/accounting/tests/test_credit_lines.py`

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Refactor: reuse invoice_date

jingcheng16 committed:
Refactor: Rename test class name

jingcheng16 committed:
Add comment to explain the test

jingcheng16 committed:
Refactor: Move the two plan to setupClass

jingcheng16 committed:
Refactor: extract code to be a helper function

jingcheng16 committed:
Refactor: rename months_from_date to get_first_day_of_months_later

The function doesn't just add months - it returns the first day of that month.

jingcheng16 committed:
Refactor: make the date logic clearer

jingcheng16 committed:
Refactor: Simplify logic to get credits for line items

Forward both `feature_type` and `is_product` down. The lower-level methods `get_credits_by_subscription_and_features` and `get_credits_for_account` will handle the checks

jingcheng16 committed:
Refactor: extract logic of generating user history

jingcheng16 committed:
Return all credits on subscription level

jingcheng16 committed:
Add new tests for product type credit and User feature type credit

jingcheng16 committed:
Make sure line item credits check include credits transferred to other subscriptions

Using the same logic for ANY type credit implemented in https://github.com/dimagi/commcare-hq/pull/26831.

jingcheng16 committed:
Rename test so it is clear that the three tests under this class if for ANY type, Product type, Feature type respectively

jingcheng16 committed:
Refactor: DRY the code

jingcheng16 committed:
No users are created in this test class, so no need to delete them

nospame commented on corehq/apps/accounting/utils/__init__.py:
```diff
@@ -34,7 +34,7 @@ def log_accounting_info(message):
 
 
-def months_from_date(reference_date, months_from_date):
+def get_first_day_of_months_later(reference_date, months_from_date):
```
`get_first_day_x_months_later`?

jingcheng16 committed:
call super()

jingcheng16 committed:
Update comments

Although Standard Plan is terminated at 9/10, when calculating prorated rate, the end date is exclusive, so use 9/9 to avoid confusion

jingcheng16 committed:
Rename function

jingcheng16 commented on corehq/apps/accounting/tests/test_credit_lines.py:
```diff
@@ -552,4 +553,11 @@ def test_subscription_credits_transfer_in_invoice(self):
+        # Credit calculation breakdown:
+        # - Initial credit: $5000.00
+        # - Standard plan (9/1-9/10): $300/month × (9/30) = $90.00
+        # - Pro plan (9/10-9/30): $600/month × (21/30) = $420.00
```
cddc3967bc61e7e89b07ee4dd5b7e942ab8162ad
